### PR TITLE
Added fix to support spaces between the method name and its `()` to prevent a false positive MCTME001 error

### LINKED ISSUE #1476: [BUG] Having a space between the UseMauiCommunityToolkitMediaElement and the `(` throws the MCTME001 error
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Did you read the "Reporting a bug" section on Contributing file?

- [X] I have read the "Reporting a bug" section on Contributing file: https://github.com/CommunityToolkit/Maui/blob/main/CONTRIBUTING.md#reporting-a-bug

### Current Behavior

I have a .editorconfig that adds a space between the method name and its `()` by default. Because of this, every time the code is formatted, it adds a space between the `UseMauiCommunityToolkitMediaElement` method name and its `()`, causing the compiler/analyzer to throw the error:

```
`.UseMauiCommunityToolkitMediaElement()` must be chained to `.UseMauiApp<T>()`(MCTME001)
```

_PS: Why I add a space between the name and its `()`? I've always used the [Mono Coding Guidelines](https://www.mono-project.com/community/contributing/coding-guidelines/)_

### Expected Behavior

Be able to compile

### Steps To Reproduce

1. Open the solution
2. Compile

### Link to public reproduction project repository

https://github.com/SotoiGhost/MauiSandbox/tree/MCT-issue

### Environment

```markdown
- .NET MAUI CommunityToolkit: 2.0.0
- OS: macOS Sonoma 14.0
- .NET MAUI: 8.0.100-rc.2.23502.2
```

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by SotoiGhost**
### Description of Change ###

Added fix to support spaces between the method name and its `()` to prevent a false positive MCTME001 error

 ### Linked Issues ###
 

 - Fixes #1476

 ### PR Checklist ###
 
 - [x] Has a linked Issue, and the Issue has been `approved`(bug) or `Championed` (feature/proposal)
 - [ ] Has tests (if omitted, state reason in description)
 - [ ] Has samples (if omitted, state reason in description)
 - [x] Rebased on top of `main` at time of PR
 - [x] Changes adhere to [coding standard](https://github.com/CommunityToolkit/Maui/blob/main/CONTRIBUTING.md#contributing-code---best-practices)
 - [ ] Documentation created or updated: https://github.com/MicrosoftDocs/CommunityToolkit/pulls 

 ### Additional information ###

### CODE REVIEW

**Review by brminnick — CHANGES_REQUESTED** [review]
Thanks for the quick turnaround on the PR @SotoiGhost! 

Could you please refactor this to use [`[GeneratedRegex]`, also known as "Compiled Regular Expressions"](https://learn.microsoft.com/en-us/dotnet/standard/base-types/regular-expression-source-generators#compiled-regular-expressions)? 

This'll help ensure we don't introduce any performance hits using RegEx.

Here's an example of how we are using `GeneratedRegex` in the toolkit: https://github.com/CommunityToolkit/Maui/blob/b430ba6a3ebcbca9f004e2430871543e4203551c/src/CommunityToolkit.Maui/Behaviors/Validators/EmailValidationBehavior.shared.cs#L21

**Review by pictos — COMMENTED** [review]
Instead of using regex, can you use the Roslyn APIs to remove any trivia, like whitespaces?

**Comment by SotoiGhost** [comment]
@brminnick I was trying to add the `[GeneratedRegex]` but I realized that the Analyzer project is a `netstandard2.0` and the attribute is only available on `.net7` and `.net8`.

image

@pictos I don't have experience with the Roslyn APIs but if you can point me into the right direction, I can make the changes, meanwhile, I'm trying to add unit testing. It's always good to learn new things! 😃

**Comment by pictos** [comment]
@SotoiGhost I don't have EOT to look into that... To not block this PR, we can merge it and in the future, I'll find time to investigate removing the regex.

**Comment by bijington** [comment]
We won't merge this until we have proven the Regex doesn't impact build performance. I have it on my list to look at 👍

**Comment by vhugogarcia** [comment]
> We won't merge this until we have proven the Regex doesn't impact build performance. I have it on my list to look at 👍

.NET 8 includes a number of Regex performance improvements, including support for compiling regular expressions more efficiently and support for executing regular expressions more efficiently.

By using this pattern: `@"\.UseMauiCommunityToolkit\s*\(\)"` . This pattern breaks down as follows:

\.: Escapes the dot since it’s a special character in Regex, ensuring it matches a literal dot.
UseMauiCommunityToolkit: Matches the literal text “UseMauiCommunityToolkit”.
\s*: Matches any whitespace character (like space, tab, etc.) zero or more times. This accounts for both no space and the space in “UseMauiCommunityToolkit ()”.
\(\): Matches the literal parentheses.

This Regex is **performant** because it avoids unnecessary backtracking and matches both text options efficiently. Here’s how you might use it in code:

```csharp
using System.Text.RegularExpressions;

string pattern = @"\.UseMauiCommunityToolkit\s*\(\)";
Regex regex = new Regex(pattern, RegexOptions.Compiled);

bool isValidOption1 = regex.IsMatch(".UseMauiCommunityToolkit()");
bool isValidOption2 = regex.IsMatch(".UseMauiCommunityToolkit ()");
```

This code snippet creates a compiled Regex object and checks if the two text options match the pattern. Both isValidOption1 and isValidOption2 should evaluate to true if the text options are as provided. The RegexOptions.Compiled option is set to this pattern since it will be used frequently, to further improve performance.

What do you all think? cc: @SotoiGhost 

I just wanted to share my grain of salt on this PR. 😃

File: samples/CommunityToolkit.Maui.Sample/MauiProgram.cs [hunk 1]
```diff
@@ -51,12 +50,12 @@ public static MauiApp CreateMauiApp()
 	{
 		var builder = MauiApp.CreateBuilder()
 #if DEBUG
-								.UseMauiCommunityToolkit(options =>
+								.UseMauiCommunityToolkit(static options =>
```
Comment (SotoiGhost) [inline samples/CommunityToolkit.Maui.Sample/MauiProgram.cs:53]: Hi @brminnick! Thanks for looking into this! Can you expand the use of `static` here, please? First time seeing this in C#. I tried to search for a doc related to this feature but without a luck. Thanks in advance!

File: src/CommunityToolkit.Maui.Analyzers/UseCommunityToolkitInitializationAnalyzer.cs [hunk 2]
```diff
@@ -33,36 +34,25 @@ static void AnalyzeNode(SyntaxNodeAnalysisContext context)
...
 	{
-		foreach (var method in root.DescendantNodes().OfType<MethodDeclarationSyntax>())
+		var stringBuilder = new StringBuilder();
+		foreach (var methodDeclaration in root.DescendantNodes().OfType<MethodDeclarationSyntax>())
 		{
-			if (method.DescendantNodes().OfType<ExpressionStatementSyntax>().Any(x => x.DescendantNodes().Any(x => x.ToString().Contains(".UseMauiCommunityToolkit("))))
-			{
-				return true;
-			}
+			stringBuilder.Append(methodDeclaration.ToString().Where(static c => !char.IsWhiteSpace(c)).ToArray());
```
[the remark this answers is not in the document]
Comment (brminnick) [reply src/CommunityToolkit.Maui.Analyzers/UseCommunityToolkitInitializationAnalyzer.cs:52]: Yes, this is a workaround. 

The existing implementation is a workaround.

Roslyn Analyzers are very difficult to build; the documentation is very light. 

If you're able to implement the Analyzer using the syntax tree, I would very much appreciate your support.

[the remark this answers is not in the document]
Comment (brminnick) [reply src/CommunityToolkit.Maui.Analyzers/UseCommunityToolkitInitializationAnalyzer.cs:52]: Awesome!! I'll add some unit tests to this PR now that use [`Microsoft.CodeAnalysis.Testing`](https://github.com/dotnet/roslyn-sdk/blob/main/src/Microsoft.CodeAnalysis.Testing/README.md)

[the remark this answers is not in the document]
Comment (brminnick) [reply src/CommunityToolkit.Maui.Analyzers/UseCommunityToolkitInitializationAnalyzer.cs]: @MartyIX Thanks so much for helping me get started on the Unit Tests! I was able to leverage your code to add a few Unit Tests to this PR that verify the happy path. 

I also updated the logic of the analyzer to use `NormalizeWhitespace` and verified it works by adding additional Unit Tests specifically testing for more whitespace.

Do you want to try to optimize the analyzers to use fewer strings? I'm not sure exactly how to accomplish that.

The next step is to verify the un-happy path (eg when `.UseMauiCommunityToolkit()` is missing), but before doing that, I wanted to see if you'd like to give it a shot!

**Review by brminnick — APPROVED** [review]
I think we're good to go! 

I've updated the Analyzers to ignore whitespace while improving performance, and I've added comprehensive Unit Test suites to verify the new changes 🚀
